File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<queue>
#include<iostream>
using namespace std;

int dx[4]={0,0,1,-1};
int dy[4]={1,-1,0,0};
int visited[101][101]={0,};
int solution(vector<vector<int> > maps)
{
    int cnt=0;
    int answer = 0;
    int n=maps.size();
    int m=maps[0].size();
    for(int i=0;i<n;i++){
        for(int j=0;j<m;j++){
            if(maps[i][j]==1 && visited[i][j]==0 && cnt==0){
                cnt++;
                queue<pair<int,int>> q;
                q.push({i,j});
                visited[i][j]=1;
                maps[i][j]
=1;                
                while(!q.empty()){
                    int front=q.front().first;
                    int back=q.front().second;
                    q.pop();
                    
                    for(int k=0;k<4;k++){
                        int nx=front+dx[k];
                        int ny=back+dy[k];
                        
                        if(nx>=0 && ny>=0 && nx<n && ny<m && visited[nx][ny]==0 && maps[nx][ny]==1){
                            q.push({nx,ny});
                            maps[nx][ny]=maps[front][back]+1;
                            visited[nx][ny]=1;
                        }
                    }
                }
            }
        }
    }
    
    
    if(visited[n-1][m-1]==1) return maps[n-1][m-1];
    else return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    static int[] dx = {1,-1,0,0};
    static int[] dy = {0,0,1,-1};
    static boolean[][] visited;
    public int solution(int[][] maps) {
        int answer = 0;
        visited = new boolean[maps.length][maps[0].length];
        Queue<Pair> q = new LinkedList<>();
        visited[0][0] = true;
        q.add(new Pair(0, 0));
        while(!q.isEmpty()){
            Pair p = q.poll();
            int x = p.x;
            int y = p.y;
            for(int i=0;i<4;i++){
                int nx = x+dx[i];
                int ny = y+dy[i];
                if(nx>=0 && nx<maps.length && ny>=0 && ny<maps[0].length && !visited[nx][ny] && maps[nx][ny] != 0){
                    maps[nx][ny] = maps[x][y] + 1;
                    q.add(new Pair(nx, ny));
                    visited[nx][ny] = true;
                }
            }
        }
        if(visited[maps.length-1][maps[0].length-1]) answer=maps[maps.length-1][maps[0].length-1];
        else answer=-1;
        return answer;
    }
    
    public static class Pair{
        int x;
        int y;
        Pair(int x, int y){
            this.x=x;
            this.y=y;
        }
    }
}